sockeye/encoder.py

```diff
@@ -262,36 +336,202 @@ def encode(self,
+        :param positions: (batch_size,)
+        :param data: (batch_size, num_embed)
+        :return: (batch_size, num_embed)
+        """
+        pass
+
+
+class AddSinCosPositionalEmbeddings(PositionalEncoder):
+    """
+    Takes an encoded sequence and adds fixed positional embeddingsas in Vaswani et al, 2017 to it.
+
+    :param num_embed: Embedding size.
+    :param max_seq_len: Maximum sequence length.
+    :param prefix: Name prefix for symbols of this encoder.
+    """
+
+    def __init__(self,
+                 num_embed: int,
+                 prefix: str) -> None:
+        utils.check_condition(num_embed % 2 == 0, "Positional embeddings require an even embedding size it "
+                              "is however %d." % num_embed)
+        self.num_embed = num_embed
+        self.prefix = prefix
+
+    def encode(self,
+               data: mx.sym.Symbol,
+               data_length: Optional[mx.sym.Symbol],
+               seq_len: int) -> Tuple[mx.sym.Symbol, mx.sym.Symbol, int]:
         """
-        Returns symbol initialized with positional encodings as in Vaswani et al.
+        :param data: (batch_size, source_seq_len, num_embed)
+        :param data_length: (batch_size,)
+        :param seq_len: sequence length.
+        :return: (batch_size, source_seq_len, num_embed)
+        """
+        embedding = mx.sym.broadcast_add(data,
+                                         mx.sym.BlockGrad(mx.symbol.Custom(length=seq_len,
+                                                                           depth=self.num_embed,
+                                                                           name="%spositional_encodings" % self.prefix,
+                                                                           op_type='positional_encodings')))
+        return embedding, data_length, seq_len
 
-        :param length: Maximum sequence length
-        :param depth: Depth.
-        :param name: Symbol name.
-        :return: Symbol(1, length, depth)
+    def encode_positions(self,
+                         positions: mx.sym.Symbol,
+                         data: mx.sym.Symbol) -> mx.sym.Symbol:
+        """
+        :param positions: (batch_size,)
+        :param data: (batch_size, num_embed)
+        :return: (batch_size, num_embed)
         """
-        return mx.sym.BlockGrad(mx.symbol.Custom(length=length, depth=depth, name=name,
-                                                 op_type='positional_encodings'))
+        # (batch_size, 1)
+        positions = mx.sym.expand_dims(positions, axis=1)
+        # (num_embed,)
+        channels = mx.sym.arange(0, self.num_embed // 2)
+        # (1, num_embed,)
+        scaling = mx.sym.expand_dims(1. / mx.sym.pow(10000, (2 * channels) / self.num_embed), axis=0)
+
+        # (batch_size, num_embed/2)
+        scaled_positions = mx.sym.dot(positions, scaling)
+
+        sin = mx.sym.sin(scaled_positions)
+        cos = mx.sym.cos(scaled_positions)
```

Comment:
This will be recomputed for each batch, right? Can we compute it in the constructor and reuse it?

Reply:
it will be recomputed, but positions will also be different here. Note that this only affects inference, as during training we still use the custom operator. I added a test to make sure both compute the same values.